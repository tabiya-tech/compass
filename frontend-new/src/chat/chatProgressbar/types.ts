export enum ConversationPhase {
  INITIALIZING = "INITIALIZING",
  INTRO = "INTRO",
  COLLECT_EXPERIENCES = "COLLECT_EXPERIENCES",
  DIVE_IN = "DIVE_IN",
  ENDED = "ENDED",
  UNKNOWN = "UNKNOWN"
}

export type CurrentPhase = {
  percentage: number
  phase: ConversationPhase
}

export const defaultCurrentPhase = {
  phase: ConversationPhase.INITIALIZING,
  percentage: 0
}
