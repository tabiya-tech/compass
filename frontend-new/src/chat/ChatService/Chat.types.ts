export interface LLMStats {
  error: string;
  prompt_token_count: number;
  response_token_count: number;
  response_time_in_sec: number;
}

export interface AgentOutput {
  message_for_user: string;
  finished: boolean;
  agent_type: string;
  reasoning: string;
  agent_response_time_in_sec: number;
  llm_stats: LLMStats[];
}

export interface Turn {
  index: number;
  input: {
    message: string;
  };
  output: AgentOutput;
}

export interface ConversationHistory {
  turns: Turn[];
}

export interface ConversationContext {
  all_history: ConversationHistory;
  history: ConversationHistory;
  summary: string;
}

export interface LastMessage {
  message_for_user: string;
  finished: boolean;
  agent_type: string;
  reasoning: string;
  agent_response_time_in_sec: number;
  llm_stats: LLMStats[];
}

export interface RootObject {
  last: LastMessage;
  conversation_context: ConversationContext;
}

export type INewChatSpecification = {
  user_id: string;
};

export type IMessageSpecification = {
  user_id: string;
  message: string;
};
