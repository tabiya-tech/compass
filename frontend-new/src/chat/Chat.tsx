import React, { useCallback, useEffect, useState, useRef } from "react";
import ChatService from "src/chat/ChatService/ChatService";
import ChatList from "src/chat/ChatList/ChatList";
import { IChatMessage } from "./Chat.types";
import { generateCompassMessage, generateUserMessage } from "./util";
import { DEFAULT_SNACKBAR_AUTO_HIDE_DURATION, useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";
import { Box, Button, ButtonProps } from "@mui/material";
import ChatHeader from "./ChatHeader/ChatHeader";
import ChatMessageField from "./ChatMessageField/ChatMessageField";
import { getUserFriendlyErrorMessage, ServiceError } from "src/error/ServiceError/ServiceError";
import { writeServiceErrorToLog } from "src/error/ServiceError/logger";
import { useNavigate } from "react-router-dom";
import { routerPaths } from "src/app/routerPaths";
import { userPreferencesStateService } from "src/userPreferences/UserPreferencesProvider/UserPreferencesStateService";
import { ConversationMessage, ConversationMessageSender, ConverstaionResponse } from "./ChatService/ChatService.types";
import { Backdrop } from "src/theme/Backdrop/Backdrop";
import ExperiencesDrawer from "src/Experiences/ExperiencesDrawer";
import { Experience } from "src/Experiences/ExperienceService/Experiences.types";
import ExperienceService from "src/Experiences/ExperienceService/ExperienceService";
import UserPreferencesService from "src/userPreferences/UserPreferencesService/userPreferences.service";
import InactiveBackdrop from "src/theme/Backdrop/InactiveBackdrop";
import { logoutService } from "src/auth/services/logout/logout.service";
import ApproveModal from "src/theme/ApproveModal/ApproveModal";
import authStateService from "src/auth/AuthStateService";
import { SnackbarKey } from "notistack";

const INACTIVITY_TIMEOUT = 3 * 60 * 1000; // in milliseconds
// Set the interval to check every TIMEOUT/3,
// so in worst case scenario the inactivity backdrop will show after TIMEOUT + TIMEOUT/3 of inactivity
export const CHECK_INACTIVITY_INTERVAL = INACTIVITY_TIMEOUT + INACTIVITY_TIMEOUT / 3;
const uniqueId = "b7ea1e82-0002-432d-a768-11bdcd186e1d";
export const DATA_TEST_ID = {
  CHAT_CONTAINER: `chat-container-${uniqueId}`,
};

interface ChatProps {
  showInactiveSessionAlert?: boolean;
  showNewExperienceAlert?: boolean;
  disableInactivityCheck?: boolean;
}

const ViewExperienceAction: React.FC<Pick<ButtonProps, "onClick">> = (props) => (
  <Button color={"info"} variant={"text"} {...props}>
    View
  </Button>
);

const Chat: React.FC<ChatProps> = ({
  showInactiveSessionAlert = false,
  showNewExperienceAlert = false,
  disableInactivityCheck = false,
}) => {
  const { enqueueSnackbar, closeSnackbar } = useSnackbar();
  const [messages, setMessages] = useState<IChatMessage[]>([]);
  const [conversationCompleted, setConversationCompleted] = useState<boolean>(false);
  const [exploredExperiences, setExploredExperiences] = useState<number>(0);
  const [conversationConductedAt, setconversationConductedAt] = useState<string | null>(null);
  const [currentMessage, setCurrentMessage] = useState<string>("");
  const [isTyping, setIsTyping] = useState<boolean>(false);
  const [isDrawerOpen, setIsDrawerOpen] = useState<boolean>(false);
  const [experiences, setExperiences] = React.useState<Experience[]>([]);
  const [isLoading, setIsLoading] = React.useState<boolean>(false);
  const [isLoggingOut, setIsLoggingOut] = React.useState<boolean>(false);
  const [showBackdrop, setShowBackdrop] = useState(showInactiveSessionAlert);
  const [lastActivityTime, setLastActivityTime] = React.useState<number>(Date.now());
  const [newConversationDialog, setNewConversationDialog] = React.useState<boolean>(false);

  const navigate = useNavigate();

  const initializationRef = useRef(false);

  /**
   * adding a message to the end of the existing messages array
   * @param message
   */
  const addMessage = (message: IChatMessage) => {
    setMessages((prevMessages) => [...prevMessages, message]);
  };

  const handleOpenExperiencesDrawer = useCallback(async () => {
    setIsDrawerOpen(true);
    setIsLoading(true);
    try {
      const userPreferences = userPreferencesStateService.getUserPreferences();
      if (!userPreferences?.sessions.length) {
        throw new Error("User has no sessions");
      }
      const experienceService = ExperienceService.getInstance(userPreferences.sessions[0]);
      const data = await experienceService.getExperiences();
      setExperiences(data);
      setIsLoading(false);
    } catch (error) {
      enqueueSnackbar("Failed to retrieve experiences", { variant: "error" });
      console.error("Failed to retrieve experiences", error);
    }
  }, [enqueueSnackbar]);

  /**
   * Action handler: View experiences drawer
   * This function is called when the user clicks on the view experiences drawer notification button
   * @returns void
   *
   */
  const viewExperiencesDrawerHandler = useCallback(
    (key: SnackbarKey) => {
      closeSnackbar(key);
      handleOpenExperiencesDrawer().then();
    },
    [closeSnackbar, handleOpenExperiencesDrawer]
  );

  const showNewExperienceAlertHandler = useCallback(
    (response: ConverstaionResponse, autoHideDuration = DEFAULT_SNACKBAR_AUTO_HIDE_DURATION) => {
      setExploredExperiences(response.experiences_explored);
      const key = enqueueSnackbar(`There is a new experience in your Skills Report.`, {
        autoHideDuration,
        variant: "info",
        action: <ViewExperienceAction onClick={() => viewExperiencesDrawerHandler(key)} />,
      });
    },
    [viewExperiencesDrawerHandler, enqueueSnackbar]
  );

  /**
   *  Send a message to compass through the chat service
   *  @param userMessage - the message to send
   *  @param session_id - the session id to use to send the message
   *  @returns void
   */
  const sendMessage = useCallback(
    async (userMessage: string, session_id?: number) => {
      const sent_at = new Date().toISOString(); // Generate current sent_at
      if (userMessage) {
        // optimistically add the user's message for a more responsive feel
        const message = generateUserMessage(userMessage, sent_at);
        addMessage(message);
      }
      try {
        if (!session_id) {
          console.error("User has no sessions");
          addMessage(
            generateCompassMessage(
              "I'm sorry, I'm having trouble connecting to the server. Please try again later.",
              sent_at
            )
          );
          return;
        }
        const chatService = ChatService.getInstance(session_id);
        if (!chatService) {
          console.error("Chat service is not initialized");
          addMessage(
            generateCompassMessage(
              "I'm sorry, I'm having trouble connecting to the server. Please try again later.",
              sent_at
            )
          );
          return;
        }
        setIsTyping(true);
        const response = await chatService.sendMessage(userMessage);
        setConversationCompleted(response.conversation_completed);
        setconversationConductedAt(response.conversation_conducted_at);

        if (response.experiences_explored > exploredExperiences) {
          showNewExperienceAlertHandler(response);
          setExploredExperiences(response.experiences_explored);
        }

        response.messages.forEach((messageItem) =>
          addMessage(generateCompassMessage(messageItem.message, messageItem.sent_at))
        );
      } catch (error) {
        console.error("Failed to send message:", error);
        addMessage(
          generateCompassMessage(
            "I'm sorry, Something seems to have gone wrong on my end... Can you repeat that?",
            sent_at
          )
        );
      } finally {
        setIsTyping(false);
      }
    },
    [exploredExperiences, showNewExperienceAlertHandler]
  );

  /**
   * Log the user out and clear the preferences
   */
  const handleLogout = useCallback(async () => {
    setIsLoggingOut(true);
    try {
      // Call the logout service to handle the logout based on the current login method
      await logoutService.handleLogout();
      // clear the user from the context, and the persistent storage
      await authStateService.clearUser();
      // clear the userPreferences from the "state"
      userPreferencesStateService.clearUserPreferences();
      navigate(routerPaths.LOGIN, { replace: true });
      enqueueSnackbar("Successfully logged out.", { variant: "success" });
    } catch (error) {
      setIsLoggingOut(false);
      console.error("Failed to logout", error);
      enqueueSnackbar(`Failed to logout: ${error}`, { variant: "error" });
    } finally {
      setIsLoggingOut(false);
    }
  }, [enqueueSnackbar, navigate]);

  const initializeChat = useCallback(
    async (session_id?: number) => {
      if (!session_id) {
        console.error("User has no sessions");
        addMessage(
          generateCompassMessage(
            "I'm sorry, Something seems to have gone wrong on my end... Can you try again?",
            new Date().toISOString()
          )
        );
        return;
      }

      const chatService = ChatService.getInstance(session_id);
      if (!chatService) {
        console.error("Chat service is not initialized");
        addMessage(
          generateCompassMessage(
            "I'm sorry, Something seems to have gone wrong on my end... Can you try again?",
            new Date().toISOString()
          )
        );
        return;
      }

      try {
        setIsTyping(true);
        const history = await chatService.getChatHistory();

        if (history.experiences_explored && history.experiences_explored > exploredExperiences) {
          setExploredExperiences(history.experiences_explored);
        }

        setConversationCompleted(history.conversation_completed);
        setconversationConductedAt(history.conversation_conducted_at);

        setMessages(
          history.messages.map((message: ConversationMessage) =>
            message.sender === ConversationMessageSender.USER
              ? generateUserMessage(message.message, message.sent_at)
              : generateCompassMessage(message.message, message.sent_at)
          )
        );

        if (!history.messages.length) {
          await sendMessage("", session_id);
        }
      } catch (e) {
        if (e instanceof ServiceError) {
          writeServiceErrorToLog(e, console.error);
        } else {
          console.error("Failed to initialize chat", e);
        }
        const errorMessage = getUserFriendlyErrorMessage(e as Error);
        enqueueSnackbar(errorMessage, { variant: "error" });
      } finally {
        setIsTyping(false);
      }
    },
    [enqueueSnackbar, sendMessage, exploredExperiences]
  );

  useEffect(() => {
    const userPreferences = userPreferencesStateService.getUserPreferences();
    if (!userPreferences || initializationRef.current) {
      return;
    }
    if (userPreferences.sessions.length) {
      initializeChat(userPreferences.sessions[0]).then(() => (initializationRef.current = true));
    }
  }, [initializeChat]);

  const handleSend = useCallback(async () => {
    if (currentMessage.trim()) {
      setCurrentMessage("");
      const userPreferences = userPreferencesStateService.getUserPreferences();
      if (!userPreferences?.sessions.length) {
        console.error("User has no sessions");
        enqueueSnackbar("Failed to send message", { variant: "error" });
        return;
      }
      await sendMessage(currentMessage, userPreferences.sessions[0]);
    }
  }, [currentMessage, sendMessage, enqueueSnackbar]);

  const startNewConversation = useCallback(async () => {
    try {
      const user = authStateService.getUser();
      if (!user?.id) return;

      const preferencesService = new UserPreferencesService();

      let user_preferences = await preferencesService.getNewSession(user?.id);

      userPreferencesStateService.setUserPreferences(user_preferences);

      enqueueSnackbar("New conversation started", { variant: "success" });

      // Initialize the new conversation with the new session id usually at the top of the sessions array
      await initializeChat(user_preferences.sessions[0]);
    } catch (e) {
      console.error("Failed to start new conversation", e);
    }
  }, [initializeChat, enqueueSnackbar]);

  const handleDrawerClose = () => {
    setIsDrawerOpen(false);
  };

  // show the user backdrop when the user is inactive for 2 minutes
  useEffect(() => {
    if (disableInactivityCheck || conversationCompleted) return;

    const checkInactivity = () => {
      if (Date.now() - lastActivityTime > INACTIVITY_TIMEOUT) {
        setShowBackdrop(true);
      }
    };
    // Check for inactivity
    const interval = setInterval(checkInactivity, CHECK_INACTIVITY_INTERVAL);

    return () => clearInterval(interval);
  }, [lastActivityTime, disableInactivityCheck, conversationCompleted]);

  useEffect(() => {
    if (showNewExperienceAlert) {
      showNewExperienceAlertHandler(
        {
          messages: [],
          experiences_explored: 10,
          conversation_completed: false,
          conversation_conducted_at: "",
        },
        DEFAULT_SNACKBAR_AUTO_HIDE_DURATION * 1000
      );
    }
  }, [showNewExperienceAlert, showNewExperienceAlertHandler]);

  // Close backdrop when user interacts with the page
  useEffect(() => {
    if (disableInactivityCheck) return;

    // Reset the timer when the user interacts with the page
    const resetTimer = () => {
      setLastActivityTime(Date.now());
      setShowBackdrop(false);
    };

    const events = ["mousedown", "keydown"];
    events.forEach((event) => document.addEventListener(event, resetTimer));

    return () => events.forEach((event) => document.removeEventListener(event, resetTimer));
  }, [disableInactivityCheck]);

  const handleOpenNewConversationDialog = () => {
    setNewConversationDialog(true);
  };

  const handleCloseNewConversationDialog = () => {
    setNewConversationDialog(false);
  };

  const handleConfirmNewConversation = async () => {
    setNewConversationDialog(false);
    await startNewConversation();
  };

  return (
    <>
      {isLoggingOut ? (
        <Backdrop isShown={isLoggingOut} message={"Logging you out, wait a moment..."} />
      ) : (
        <>
          <Box
            width="100%"
            height="100%"
            display="flex"
            flexDirection="column"
            data-testid={DATA_TEST_ID.CHAT_CONTAINER}
          >
            <ChatHeader
              notifyOnLogout={handleLogout}
              startNewConversation={handleOpenNewConversationDialog}
              notifyOnExperiencesDrawerOpen={handleOpenExperiencesDrawer}
            />
            <Box sx={{ flex: 1, overflowY: "auto" }}>
              <ChatList messages={messages} isTyping={isTyping} />
            </Box>
            {showBackdrop && <InactiveBackdrop isShown={showBackdrop} />}
            <Box sx={{ flexShrink: 0 }}>
              <ChatMessageField
                handleSend={handleSend}
                aiIsTyping={isTyping}
                isChatFinished={conversationCompleted}
                message={currentMessage}
                notifyChange={setCurrentMessage}
              />
            </Box>
          </Box>
          <ExperiencesDrawer
            isOpen={isDrawerOpen}
            notifyOnClose={handleDrawerClose}
            isLoading={isLoading}
            experiences={experiences}
            conversationConductedAt={conversationConductedAt}
          />
          {newConversationDialog && (
            <ApproveModal
              isOpen={newConversationDialog}
              title="Start New Conversation?"
              content={
                <>
                  Once you start a new conversation, all messages from the current conversation will be lost forever.
                  <br />
                  <br />
                  Are you sure you want to start a new conversation?
                </>
              }
              onCancel={handleCloseNewConversationDialog}
              onApprove={handleConfirmNewConversation}
              cancelButtonText="Cancel"
              approveButtonText="Yes, I'm sure"
            />
          )}
        </>
      )}
    </>
  );
};

export default Chat;
