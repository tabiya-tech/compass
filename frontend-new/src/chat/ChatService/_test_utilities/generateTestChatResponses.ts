import { ConversationMessage, ConversationMessageSender } from "src/chat/ChatService/ChatService.types";

export const generateTestChatResponses = (): ConversationMessage[] => {
  return [
    {
      message: "Hello! How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "I can help you with that. What is your name?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
  ];
};

export const generateTestHistory = (): ConversationMessage[] => {
  return [
    {
      message: "Hello! How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "I can help you with that. What is your name?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "My name is John.",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.USER,
    },
    {
      message: "Nice to meet you, John. How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
  ];
};
