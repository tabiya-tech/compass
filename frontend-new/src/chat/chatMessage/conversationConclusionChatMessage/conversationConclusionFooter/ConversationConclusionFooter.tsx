import React, { useState, useEffect } from "react";
import { Box, Typography, useTheme } from "@mui/material";
import CustomLink from "src/theme/CustomLink/CustomLink";
import { FIXED_MESSAGES_TEXT } from "src/chat/util";
import FeedbackForm, {
  FeedbackStatus,
  FeedbackCloseEvent,
} from "src/feedback/overallFeedback/feedbackForm/FeedbackForm";
import { useChatContext } from "src/chat/ChatContext";
import UserPreferencesStateService from "src/userPreferences/UserPreferencesStateService";
import { FeedbackItem } from "src/feedback/overallFeedback/overallFeedbackService/OverallFeedback.service.types";
import { PersistentStorageService } from "src/app/PersistentStorageService/PersistentStorageService";
import authenticationStateService from "src/auth/services/AuthenticationState.service";
import AnonymousAccountConversionDialog from "src/auth/components/anonymousAccountConversionDialog/AnonymousAccountConversionDialog";
import CustomerSatisfactionRating from "src/feedback/overallFeedback/feedbackForm/components/customerSatisfactionRating/CustomerSatisfaction";

const uniqueId = "41675f8b-257c-4a63-9563-3fe9feb6a850";

export const DATA_TEST_ID = {
  CONVERSATION_CONCLUSION_FOOTER_CONTAINER: `feedback-form-button-container-${uniqueId}`,
  FEEDBACK_FORM_BUTTON: `feedback-form-button-${uniqueId}`,
  EXPERIENCES_DRAWER_BUTTON: `feedback-form-experiences-drawer-button-${uniqueId}`,
  FEEDBACK_IN_PROGRESS_BUTTON: `feedback-in-progress-button-${uniqueId}`,
  FEEDBACK_MESSAGE_TEXT: `feedback-message-text-${uniqueId}`,
  FEEDBACK_IN_PROGRESS_MESSAGE: `feedback-in-progress-message-${uniqueId}`,
  THANK_YOU_FOR_FEEDBACK_MESSAGE: `thank-you-for-feedback-message-${uniqueId}`,
  THANK_YOU_FOR_RATING_MESSAGE: `thank-you-for-rating-message${uniqueId}`,
  CREATE_ACCOUNT_LINK: `create-account-link-${uniqueId}`,
  CREATE_ACCOUNT_MESSAGE: `create-account-message-${uniqueId}`,
  VERIFICATION_REMINDER_MESSAGE: `verification-reminder-message-${uniqueId}`,
};

const ConversationConclusionFooter: React.FC = () => {
  const theme = useTheme();
  const { feedbackStatus, setFeedbackStatus, handleOpenExperiencesDrawer } = useChatContext();

  const [feedbackData, setFeedbackData] = useState<FeedbackItem[]>(PersistentStorageService.getOverallFeedback());
  const [sessionHasFeedback, setSessionHasFeedback] = useState<boolean>(
    UserPreferencesStateService.getInstance().activeSessionHasOverallFeedback()
  );
  const [isFeedbackFormOpen, setIsFeedbackFormOpen] = useState<boolean>(false);
  const [showConversionDialog, setShowConversionDialog] = useState<boolean>(false);
  const [hasSubmittedCustomerSatisfactionRating, setHasSubmittedCustomerSatisfactionRating] = useState<boolean>(
    UserPreferencesStateService.getInstance().activeSessionHasCustomerSatisfactionRating()
  );

  const user = authenticationStateService.getInstance().getUser();
  const isAnonymous = !user?.name || !user?.email;
  const { isAccountConverted, setIsAccountConverted } = useChatContext();

  // Check local storage when the form is closed to see if there is saved feedback
  useEffect(() => {
    if (!isFeedbackFormOpen) {
      const updatedFeedbackData = PersistentStorageService.getOverallFeedback();
      if (JSON.stringify(feedbackData) !== JSON.stringify(updatedFeedbackData)) {
        setFeedbackData(updatedFeedbackData);
        // If there is feedback data, set the feedback status to started
        if (updatedFeedbackData.length > 0 && feedbackStatus === FeedbackStatus.NOT_STARTED) {
          setFeedbackStatus(FeedbackStatus.STARTED);
        }
      }
    }
  }, [isFeedbackFormOpen, feedbackData, feedbackStatus, setFeedbackStatus]);

  // Update feedback status based on session feedback state
  useEffect(() => {
    if (sessionHasFeedback) {
      setFeedbackStatus(FeedbackStatus.SUBMITTED);
    } else if (feedbackData.length > 0) {
      setFeedbackStatus(FeedbackStatus.STARTED);
    }
  }, [feedbackData.length, sessionHasFeedback, setFeedbackStatus]);


  const handleFeedbackFormClose = (closeEvent: FeedbackCloseEvent) => {
    setIsFeedbackFormOpen(false);
    if (closeEvent === FeedbackCloseEvent.SUBMIT) {
      setFeedbackStatus(FeedbackStatus.SUBMITTED);
      setSessionHasFeedback(true);
    } else if (closeEvent === FeedbackCloseEvent.DISMISS) {
      if (feedbackData.length > 0) {
        setFeedbackStatus(FeedbackStatus.STARTED);
      } else {
        setFeedbackStatus(FeedbackStatus.NOT_STARTED);
      }
    }
  };

  const feedbackSubmitted = (sessionHasFeedback || feedbackStatus === FeedbackStatus.SUBMITTED);

  return (
    <>
      <Box
        display="flex"
        flexDirection="column"
        justifyContent="start"
        gap={theme.fixedSpacing(theme.tabiyaSpacing.md)}
        marginTop={theme.fixedSpacing(theme.tabiyaSpacing.md)}
        data-testid={DATA_TEST_ID.CONVERSATION_CONCLUSION_FOOTER_CONTAINER}
      >
        <Typography variant="body1">
          You can now{" "}
          <CustomLink
            onClick={handleOpenExperiencesDrawer}
            disableWhenOffline
            data-testid={DATA_TEST_ID.EXPERIENCES_DRAWER_BUTTON}
          >
            <span style={{ whiteSpace: "normal" }}>view and download your CV here</span>
          </CustomLink>
          . {/* Show anonymous user registration link if the user is anonymous and hasn't already converted */}
          {isAnonymous && !isAccountConverted && (
            <span data-testid={DATA_TEST_ID.CREATE_ACCOUNT_MESSAGE}>
              If you would like to explore your skills and experiences in the future, you can{" "}
              <CustomLink
                onClick={() => setShowConversionDialog(true)}
                disableWhenOffline
                data-testid={DATA_TEST_ID.CREATE_ACCOUNT_LINK}
              >
                Create an account
              </CustomLink>
            </span>
          )}
          {/* Show the verification reminder if the user has already converted their account */}
          {isAccountConverted && (
            <span data-testid={DATA_TEST_ID.VERIFICATION_REMINDER_MESSAGE}>
              A verification email has been sent to your email address. Please verify your account before logging in
              again.
            </span>
          )}
        </Typography>

        <Typography component={"span"} variant="body1">
          {/* Show a customer satisfaction rating if it hasn't been submitted yet */}
          {!hasSubmittedCustomerSatisfactionRating && (
            <CustomerSatisfactionRating
              notifyOnCustomerSatisfactionRatingSubmitted={() => setHasSubmittedCustomerSatisfactionRating(true)}
            />
          )}
          {/* Show a thank you message if the rating has been submitted */}
          {hasSubmittedCustomerSatisfactionRating && !feedbackSubmitted && (
            <span data-testid={DATA_TEST_ID.THANK_YOU_FOR_RATING_MESSAGE}>
              {" "}
              {FIXED_MESSAGES_TEXT.THANK_YOU_FOR_RATING}{" "}
            </span>
          )}
          {/* Show feedback form if the rating has been submitted */}
          {hasSubmittedCustomerSatisfactionRating &&
            feedbackStatus === FeedbackStatus.NOT_STARTED &&
              <span data-testid={DATA_TEST_ID.FEEDBACK_MESSAGE_TEXT}>{" "}We'd love to get more{" "}
                <CustomLink
                  onClick={() => setIsFeedbackFormOpen(true)}
                  disableWhenOffline
                  data-testid={DATA_TEST_ID.FEEDBACK_FORM_BUTTON}
                >
                feedback
              </CustomLink>{" "}
                from you. It only takes 5 minutes and helps us improve!
            </span>
          }
          {/* Show continue feedback if the status is already started */}
          {hasSubmittedCustomerSatisfactionRating && feedbackStatus === FeedbackStatus.STARTED && (
            <span data-testid={DATA_TEST_ID.FEEDBACK_IN_PROGRESS_MESSAGE}>
              Please{" "}
              <CustomLink
                onClick={() => setIsFeedbackFormOpen(true)}
                disableWhenOffline
                data-testid={DATA_TEST_ID.FEEDBACK_IN_PROGRESS_BUTTON}
              >
                complete your feedback
              </CustomLink>{" "}
              to help us improve your experience!
            </span>
          )}
          {feedbackSubmitted && hasSubmittedCustomerSatisfactionRating && (
            <span data-testid={DATA_TEST_ID.THANK_YOU_FOR_FEEDBACK_MESSAGE}>
              {FIXED_MESSAGES_TEXT.THANK_YOU_FOR_FEEDBACK}
            </span>
          )}
        </Typography>
      </Box>
      <FeedbackForm isOpen={isFeedbackFormOpen} notifyOnClose={handleFeedbackFormClose} />
      <AnonymousAccountConversionDialog
        isOpen={showConversionDialog}
        onClose={() => setShowConversionDialog(false)}
        onSuccess={() => {
          setIsAccountConverted(true);
        }}
      />
    </>
  );
};

export default ConversationConclusionFooter;

