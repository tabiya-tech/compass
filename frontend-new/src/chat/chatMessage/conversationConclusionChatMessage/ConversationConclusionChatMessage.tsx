import React from "react";
import { IChatMessage } from "src/chat/Chat.types";
import { MessageContainer } from "src/chat/chatMessage/basicChatMessage/BasicChatMessage";
import ConversationConclusionFooter from "src/chat/chatMessage/conversationConclusionChatMessage/conversationConclusionFooter/ConversationConclusionFooter";
import ChatBubble from "src/chat/chatMessage/components/chatBubble/ChatBubble";

const uniqueId = "2fbaf2ef-9eab-485a-bd28-b4a164e18b06";

export const DATA_TEST_ID = {
  CONVERSATION_CONCLUSION_CHAT_MESSAGE_CONTAINER: `conversation_conclusion_chat-message-container-${uniqueId}`,
};

interface ConversationConclusionChatMessageProps {
  chatMessage: IChatMessage;
}

const ConversationConclusionChatMessage: React.FC<ConversationConclusionChatMessageProps> = ({
  chatMessage,
}) => {
  return (
    <MessageContainer
      origin={chatMessage.sender}
      data-testid={DATA_TEST_ID.CONVERSATION_CONCLUSION_CHAT_MESSAGE_CONTAINER}
    >
      <ChatBubble message={chatMessage.message} sender={chatMessage.sender}>
        <ConversationConclusionFooter />
      </ChatBubble>
    </MessageContainer>
  );
};

export default ConversationConclusionChatMessage;
