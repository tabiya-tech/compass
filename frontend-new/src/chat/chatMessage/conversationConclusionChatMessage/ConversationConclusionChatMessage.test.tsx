// mute the console
import "src/_test_utilities/consoleMock";

import ConversationConclusionChatMessage, { DATA_TEST_ID } from "./ConversationConclusionChatMessage";
import ConversationConclusionFooter from "src/chat/chatMessage/conversationConclusionChatMessage/conversationConclusionFooter/ConversationConclusionFooter";
import {
  DATA_TEST_ID as CHAT_BUBBLE_DATA_TEST_ID,
} from "src/chat/chatMessage/components/chatBubble/ChatBubble";
import { render, screen } from "src/_test_utilities/test-utils";
import { ConversationMessageSender } from "src/chat/ChatService/ChatService.types";
import { nanoid } from "nanoid";
import { ChatMessageType, IChatMessage } from "src/chat/Chat.types";

jest.mock("src/chat/chatMessage/components/chatBubble/ChatBubble", () => {
  const originalModule = jest.requireActual("src/chat/chatMessage/components/chatBubble/ChatBubble");
  return {
    __esModule: true,
    ...originalModule,
    default: jest.fn(({children}) => <div data-testid={originalModule.DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_CONTAINER}>{children}</div>),
  };
});

jest.mock(
  "src/chat/chatMessage/conversationConclusionChatMessage/conversationConclusionFooter/ConversationConclusionFooter",
  () => {
    const originalModule = jest.requireActual(
      "src/chat/chatMessage/conversationConclusionChatMessage/conversationConclusionFooter/ConversationConclusionFooter"
    );
    return {
      __esModule: true,
      ...originalModule,
      default: jest.fn(() => (
        <div data-testid={originalModule.DATA_TEST_ID.CONVERSATION_CONCLUSION_FOOTER_CONTAINER}></div>
      )),
    };
  }
);

describe("render tests", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  })
  test("should render the Chat message", () => {
    // GIVEN a basic chat message sent at a given time
    const givenDate = new Date().toISOString();
    const givenMessage: IChatMessage = {
      message_id: nanoid(),
      sender: ConversationMessageSender.COMPASS,
      message: "Thanks for having a conversation with me.",
      sent_at: givenDate,
      type: ChatMessageType.CONVERSATION_CONCLUSION, // This component is designed for use with the Conversation conclusion chat type
      reaction: null, // Conversation Conclusion message cant have a reaction
    };

    // WHEN the conversation conclusion chat message is rendered
    render(
      <ConversationConclusionChatMessage
        chatMessage={givenMessage}
      />
    );

    // THEN expect the message container to be visible
    expect(screen.getByTestId(DATA_TEST_ID.CONVERSATION_CONCLUSION_CHAT_MESSAGE_CONTAINER)).toBeInTheDocument();
    // AND expect the message bubble to be visible
    expect(screen.getByTestId(CHAT_BUBBLE_DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_CONTAINER)).toBeInTheDocument();

    // AND expect the conversation conclusion footer to be visible
    expect(ConversationConclusionFooter).toHaveBeenCalledTimes(1);

    // AND expect the component to match the snapshot
    expect(screen.getByTestId(DATA_TEST_ID.CONVERSATION_CONCLUSION_CHAT_MESSAGE_CONTAINER)).toMatchSnapshot();
  });

  test("should render with feedback in progress", () => {
    // GIVEN a basic chat message sent at a given time
    const givenDate = new Date().toISOString();
    const givenMessage: IChatMessage = {
      message_id: nanoid(),
      sender: ConversationMessageSender.COMPASS,
      message: "Thanks for having a conversation with me.",
      sent_at: givenDate,
      type: ChatMessageType.CONVERSATION_CONCLUSION,
      reaction: null
    };

    // WHEN the conversation conclusion chat message is rendered
    render(
      <ConversationConclusionChatMessage
        chatMessage={givenMessage}
      />
    );

    // THEN expect the conversation conclusion footer to be visible
    expect(ConversationConclusionFooter).toHaveBeenCalledTimes(1);
  });

  test("should render with feedback submitted", () => {
    // GIVEN a basic chat message sent at a given time
    const givenDate = new Date().toISOString();
    const givenMessage: IChatMessage = {
      message_id: nanoid(),
      sender: ConversationMessageSender.COMPASS,
      message: "Thanks for having a conversation with me.",
      sent_at: givenDate,
      type: ChatMessageType.CONVERSATION_CONCLUSION,
      reaction: null
    };

    // WHEN the conversation conclusion chat message is rendered
    render(
        <ConversationConclusionChatMessage
          chatMessage={givenMessage}
        />
    );

    // THEN expect the conversation conclusion footer to be visible
    expect(ConversationConclusionFooter).toHaveBeenCalledTimes(1);
  });
});
