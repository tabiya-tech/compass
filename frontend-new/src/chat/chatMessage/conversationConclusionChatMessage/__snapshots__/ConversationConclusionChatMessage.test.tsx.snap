// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render tests should render the Chat message 1`] = `
<div
  class="MuiBox-root css-1c8zvho"
  data-testid="conversation_conclusion_chat-message-container-2fbaf2ef-9eab-485a-bd28-b4a164e18b06"
  origin="COMPASS"
>
  <div
    data-testid="chat-message-bubble-container-6e685eeb-2b54-432a-8b66-8a81633b3981"
  >
    <div
      data-testid="feedback-form-button-container-41675f8b-257c-4a63-9563-3fe9feb6a850"
    />
  </div>
</div>
`;
